### PR TITLE
Extended bridge testcases

### PR DESCRIPTION
**Pull request by Dinaamagdy**

### CODE REVIEW

File: functional_testing/Grid_API_Testing/api_testing/testcases/extend_tests/test02_bridges.py [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+        time.sleep(3)
+
+        self.lg.info('Delete bridge (B1), should succeed..')
+        response = self.bridges_api.delete_nodes_bridges_bridgeid(self.nodeid, self.bridge_name)
+        self.assertEqual(response.status_code, 204, response.content)
+
+        self.lg.info('Create bridge (B2) with same name for (B1),should succeed')
+        response = self.bridges_api.delete_nodes_bridges_bridgeid(self.nodeid, self.bridge_name)
+        self.assertEqual(response.status_code, 201, response.content)
+        self.createdcontainer.append({"node": self.nodeid, "name": self.bridge_name})
```
Comment (ahmedelsayed-93) [inline functional_testing/Grid_API_Testing/api_testing/testcases/extend_tests/test02_bridges.py]: ```createdbridges``` not ```createdcontainer```

File: functional_testing/Grid_API_Testing/api_testing/testcases/extend_tests/test02_bridges.py [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+
+    def test002_create_bridge_with_nat(self):
+        """ GAT-002
+        *Create bridge with nat options *
+
+        **Test Scenario:**
+
+        #. Create bridge (B0) with false in nat option,should succeed.
+        #. Create (C0) with (B0) bridge,should succeed.
```
Comment (ahmedelsayed-93) [inline functional_testing/Grid_API_Testing/api_testing/testcases/extend_tests/test02_bridges.py]: add  ** container** before (C0)

File: functional_testing/Grid_API_Testing/api_testing/testcases/extend_tests/test02_bridges.py [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+
+        """
+        nat_options = [False, True]
+        for i, nat in enumerate(nat_options):
+            B_name = self.rand_str_chars_only()
+            B_body = {"name": B_name,
+                      "hwaddr": self.randomMAC(),
+                      "networkMode": "dnsmasq",
+                      "nat": nat,
+                      "setting": {"cidr": "217.102.2.0/8",
```
Comment (ahmedelsayed-93) [inline functional_testing/Grid_API_Testing/api_testing/testcases/extend_tests/test02_bridges.py]: bridge ip is invalid change cidr value

File: functional_testing/Grid_API_Testing/api_testing/testcases/extend_tests/test02_bridges.py [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+        self.lg.info("Create bridge (B1) with wrong hardware address, should fail.")
+        hardwareaddr = self.rand_str()
+        B1_name = self.rand_str()
+        self.bridge_body["hwaddr"] = hardwareaddr
+        self.bridge_body["name"] = B1_name
+        response = self.bridges_api.post_nodes_bridges(self.nodeid, self.bridge_body)
+        self.createdbridges.append({"node": self.nodeid, "name": self.bridge_name})
+        self.assertEqual(response.status_code, 400, response.content)
+
+    def test004_create_bridge_with_static_networkMode(self):
```
Comment (islamTaha12) [inline functional_testing/Grid_API_Testing/api_testing/testcases/extend_tests/test02_bridges.py]: Divide this tes tcase to 5 test cases cause there is no need to  compress those 5 scenarios in the same test case

File: functional_testing/Grid_API_Testing/api_testing/testcases/extend_tests/test02_bridges.py [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+        B3_name = self.rand_str()
+        self.bridge_body["name"] = B2_name
+        self.bridge_body["networkMode"] = "static"
+        self.bridge_body["setting"] = {}
+        response = self.bridges_api.post_nodes_bridges(self.nodeid,self.bridge_body)
+        self.createdbridges.append({"node": self.nodeid, "name": B3_name})
+        self.assertEqual(response.status_code, 400, response.content)
+
+
+    def test005_create_bridge_with_dnsmasq_networkMode(self):
```
Comment (islamTaha12) [inline functional_testing/Grid_API_Testing/api_testing/testcases/extend_tests/test02_bridges.py]: Divide this test case

File: functional_testing/Grid_API_Testing/api_testing/testcases/extend_tests/test02_bridges.py [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+        """ GAT-005
```
Comment (islamTaha12) [inline functional_testing/Grid_API_Testing/api_testing/testcases/extend_tests/test02_bridges.py]: update this index

File: functional_testing/Grid_API_Testing/api_testing/testcases/extend_tests/test02_bridges.py [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+        #. Create bridge (B3) with dnsmasq network mode and empty setting,should fail.
+        #. Create bridge (B4) with dnsmasq network and empty start and end values, should fail.
+
+        """
+
+        cidr_address = "205.102.2.1/8"
+        start = "205.102.3.2"
+        end = "205.102.3.3"
+        self.bridge_body["networkMode"] = "dnsmasq"
+        self.bridge_body["setting"] = {"cidr": "%s"%cidr_address, "start": start, "end": end}
```
Comment (islamTaha12) [inline functional_testing/Grid_API_Testing/api_testing/testcases/extend_tests/test02_bridges.py]: change "%s"

File: functional_testing/Grid_API_Testing/api_testing/testcases/extend_tests/test02_bridges.py [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+
+        self.lg.info("Check that (B0)bridge took given cidr address, should succeed.")
+        response = self.nodes_api.get_nodes_nodeid_nics(self.nodeid)
+        self.assertEqual(response.status_code, 200)
+        nic = [x for x in response.json() if x["name"] == self.bridge_name][0]
+        self.assertIn(cidr_address, nic["addrs"])
+
+        self.lg.info(" Create bridge (B1) overlapping with (B0) address,shoud fail.")
+        B1_name = self.rand_str()
+        cidr_address = "205.103.2.0/8"
```
Comment (islamTaha12) [inline functional_testing/Grid_API_Testing/api_testing/testcases/extend_tests/test02_bridges.py]: u should not set 2.0/8

File: functional_testing/Grid_API_Testing/api_testing/testcases/extend_tests/test02_bridges.py [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+        start = "205.103.3.1"
+        end = "205.103.3.1"
```
Comment (islamTaha12) [inline functional_testing/Grid_API_Testing/api_testing/testcases/extend_tests/test02_bridges.py]: there is no ip here ... start and end are the same

Comment (Dinaamagdy) [inline functional_testing/Grid_API_Testing/api_testing/testcases/extend_tests/test02_bridges.py]: done

File: functional_testing/Grid_API_Testing/api_testing/testcases/extend_tests/test02_bridges.py (hunk 8 above)
Comment (Dinaamagdy) [inline functional_testing/Grid_API_Testing/api_testing/testcases/extend_tests/test02_bridges.py]: done

File: functional_testing/Grid_API_Testing/api_testing/testcases/extend_tests/test02_bridges.py [hunk 10]
```diff
@@ -0,0 +1,294 @@
...
+        self.createdbridges = []
+
+    def setUp(self):
+        super(TestBridgesAPI, self).setUp()
+
+        self.lg.info('Get random nodid (N0)')
+        self.nodeid = self.get_random_node()
+        pyclient_ip = [x['ip'] for x in self.nodes if x['id'] == self.nodeid][0]
+        self.root_url = "https://hub.gig.tech/gig-official-apps/ubuntu1604.flist"
+        self.pyclient = Client(pyclient_ip)
```
Comment (Dinaamagdy) [inline functional_testing/Grid_API_Testing/api_testing/testcases/extend_tests/test02_bridges.py]: done

File: functional_testing/Grid_API_Testing/api_testing/testcases/extend_tests/test02_bridges.py [hunk 11]
```diff
@@ -0,0 +1,294 @@
...
+                            }
+
+    def tearDown(self):
+        self.lg.info('TearDown:delete all created bridges ')
+        for bridge in self.createdbridges:
+            self.bridges_api.delete_nodes_bridges_bridgeid(bridge['node'],
+                                                           bridge['name'])
+
+    def test001_create_bridges_with_same_name(self):
+        """ GAT-001
```
Comment (Dinaamagdy) [inline functional_testing/Grid_API_Testing/api_testing/testcases/extend_tests/test02_bridges.py]: done

File: functional_testing/Grid_API_Testing/api_testing/testcases/extend_tests/test02_bridges.py [hunk 12]
```diff
@@ -0,0 +1,294 @@
...
+        *Create two bridges with same name *
+
+        **Test Scenario:**
+
+        #. Create bridge (B1) , should succeed .
+        #. Create bridge (B2) with same name for (B1),should fail.
```
Comment (Dinaamagdy) [inline functional_testing/Grid_API_Testing/api_testing/testcases/extend_tests/test02_bridges.py]: done

File: functional_testing/Grid_API_Testing/api_testing/testcases/extend_tests/test02_bridges.py [hunk 13]
```diff
@@ -0,0 +1,294 @@
...
+        response = self.bridges_api.post_nodes_bridges(self.nodeid, self.bridge_body)
+        self.assertEqual(response.status_code, 409, response.content)
+        time.sleep(3)
+
+        self.lg.info('Delete bridge (B1), should succeed..')
+        response = self.bridges_api.delete_nodes_bridges_bridgeid(self.nodeid, self.bridge_name)
+        self.assertEqual(response.status_code, 204, response.content)
+
+    def test002_create_bridge_with_nat(self):
+        """ GAT-002
```
Comment (Dinaamagdy) [inline functional_testing/Grid_API_Testing/api_testing/testcases/extend_tests/test02_bridges.py]: done

File: functional_testing/Grid_API_Testing/api_testing/testcases/extend_tests/test02_bridges.py [hunk 14]
```diff
@@ -0,0 +1,294 @@
...
+        #. Create bridge (B0), should succeed.
+        #. Check that (B0)bridge took given cidr address, should succeed.
+        #. Create bridge(B1)with same cidr as (B0),should fail.
+        #. Create bridge (B2) with invalid cidr address, should fail.
+        #. Create bridge (B3) with static network mode and empty cidr value,should fail.
+
+        """
+        cidr_address = "130.111.3.1/8"
+        self.bridge_body["networkMode"] = "static"
+        self.bridge_body["setting"] = {"cidr": "%s"%cidr_address}
```
Comment (Dinaamagdy) [inline functional_testing/Grid_API_Testing/api_testing/testcases/extend_tests/test02_bridges.py]: done

File: functional_testing/Grid_API_Testing/api_testing/testcases/extend_tests/test02_bridges.py (hunk 7 above)
Comment (Dinaamagdy) [inline functional_testing/Grid_API_Testing/api_testing/testcases/extend_tests/test02_bridges.py]: done

File: functional_testing/Grid_API_Testing/api_testing/testcases/extend_tests/test02_bridges.py (hunk 4 above)
Comment (Dinaamagdy) [inline functional_testing/Grid_API_Testing/api_testing/testcases/extend_tests/test02_bridges.py]: done

**Pull request merged**